### ISSUE TITLE
Closing Window Animation leaves artifacts.

### ISSUE THREAD

**Issue by Moshe-Shelomov** [allusive-dev/picom-allusive#3]
When I switch workspaces, randomly appears a little square on the screen (sometimes more than one).
if I switch back and forth it eventually disappears but it's quite irritating.
Visually it appears as if programs are zooming out but instead of disappearing they are oddly minimized

I use EndeavourOS and Qtile
Let me know if I need to provide any further info

in the attached zip is my picom config and screen recording showing the glitch

[Archive.zip](https://github.com/allusive-dev/picom-allusive/files/12666623/Archive.zip)

**Comment by allusive-dev** [comment]
Hello Moshe!
I am sorry this is happening to you, I personally haven't experienced this before but I will try my best to assist.

After using your config I cannot seem to replicate this issue.
My Specs:
Arch Linux
NVIDIA RTX 3060 (`nvidia-open`)
Intel i5-12600

Would you be able to share your hardware specs as well as what graphics driver you are using along with your monitor resolutions.

**Comment by Moshe-Shelomov** [comment]
> Hello Moshe! I am sorry this is happening to you, I personally haven't experienced this before but I will try my best to assist.
> 
> After using your config I cannot seem to replicate this issue. My Specs: Arch Linux NVIDIA RTX 3060 (`nvidia-open`) Intel i5-12600 *Xmonad
> 
> Would you be able to share your hardware specs as well as what graphics driver you are using along with your monitor resolutions.

I'm on a Alienware (Dell) laptop ,  model: "17 R3"
CPU - Intel i7 6700HQ
GPU - Nvidia 970M (nvidia-dkms driver from AUR)
resolution - 1920*1080
Kernel: 6.5.3-zen1-1-zen

since it's a laptop, I'm also using "optimus manager" for switching graphics
(https://github.com/Askannz/optimus-manager)

**Comment by allusive-dev** [comment]
My first observation is that you have not set a window-unmap animation in your configuration. Try setting it to `zoom` or check the [wiki's animations page](https://github.com/allusive-dev/picom-allusive/wiki/ANIMATIONS!!!#animation-for-unmap-window).
I have an older nvidia laptop with amd iGPU so I could maybe try installing endeavourOs and testing it at some point if necessary. 

My other guess is that it could be your graphics driver. I know that the proprietary nvidia driver causes quite a couple of odd issues for picom and this could be one of them. So if you are able to, one suggestion might be disabling your iGPU from bios, not using optimus-manager and installing the `nvidia-open` package. (I have done this on my own system to fix conflicts with my iGPU)
The above option is sort of a last resort considering the bios changes. 

A few other questions. 
Does the black box appear for all windows or only affect some?
If the box appears for all windows then does the color of the box reflect the coloring of the window causing it?

**Comment by Moshe-Shelomov** [comment]
From my observation, the box appears for any window and it does reflect the color of the window
So when it is caused by librewolf for example, it has the color of the current page opened in it, etc.

**Comment by Moshe-Shelomov** [comment]
I've just added to config following lines:

animation-for-open-window = "zoom";
animation-for-unmap-window = "zoom";
animation-for-workspace-switch-in = "zoom";
animation-for-workspace-switch-out = "zoom";
animation-for-transient-window = "zoom";

...but the glitch is still there.
I'll try changing "animation-clamping" to true, since in the wiki you wrote that it's not known what it does, maybe it'll help

**Comment by draculalaa** [comment]
Same problem to me, and if I change workspace switch animation style from zoom to "slide-", it appears to be a slender rectangle, rather than a little square remains on the screen. 

I belive it's not caused by graphic driver, since I'm using a laptop with AMD 6800H and integrated graphics.

And another problem is when I switch the workspace slowly (after the animation is fully played I think) , the last focused window's image will stay in new workspace. 

Seems like all above are caused by same reason. I think they are related to WM's behavior, but my poor searching skill find nothing about that.

OS: Arch Linux x86_64
WM: bspwm
CPU: AMD Ryzen 7 6800H (16) @ 4.785 GHz
GPU: AMD Radeon 680M

[vid.zip](https://github.com/allusive-dev/picom-allusive/files/12671298/vid.zip)
[picom.conf.txt](https://github.com/allusive-dev/picom-allusive/files/12671307/picom.conf.txt)

**Comment by allusive-dev** [comment]
Hey @draculalaa and @Moshe-Shelomov.
Thank you very muhc for your contributions to helping solve this issue.
From what I can tell, windows that have been closed after being moved to a new workspace they leave "residue" in different patterns depending on which unmap animation is set. 

My first idea would be to recode the unmap animation slightly however this doesn't make sense since I and others do not experience this issue.

I agree with you now @draculalaa about it not being a graphics driver issue. Considering 2 different people with different GPU's are experiencing the issue it must be something deeper.

I am going to do some research and get back to you both asap.
Thank you for your patience and I hope we can resolve this!

**Comment by allusive-dev** [comment]
> [vid.zip](https://github.com/allusive-dev/picom-allusive/files/12671298/vid.zip) 

This video that you sent is quite odd. The shaping of the window residue is extremely odd and not logical to the unmap animation.
Another thing to figure out I guess.

**Comment by allusive-dev** [comment]
To help diagnose the issue can both @draculalaa and @Moshe-Shelomov try using this picom config and tell me whether the issue is still present.
Thanks again!

[picom.conf.txt](https://github.com/allusive-dev/picom-allusive/files/12672542/picom.conf.txt)

**Comment by Moshe-Shelomov** [comment]
yes, the issue remains with this config as well

**Comment by Moshe-Shelomov** [comment]
I switched to qtile just a few days ago and before that I was using i3, so I've decided to check whether this issue will appear there as well. After playing around with windows and workspaces and whatever in i3, there was no issue whatsoever.(with my picom config)
Perhaps there's something in my qtile config that is causing the issue, or there might be some incompatibility with qtile in general?

**Comment by allusive-dev** [comment]
Hello @Moshe-Shelomov 
Yeah well considering that the config I sent you did not fix it but you also had no issues on another WM. My best guess like you said is something to do with your WM/wm config.
I haven't used qtile before but I am surprised since it is basically just xmonad in python.

**Comment by allusive-dev** [comment]
Considering what @draculalaa said this issue may also extend to bspwm. However I can't be sure of this at the moment until they.
1. Try the config i sent
2. Try a different WM

**Comment by draculalaa** [comment]
Hi,
1. I've tried the config file @allusive-dev provided above and problem remains. 
2. And I tried i3wm with basically default config, there is no animation glitch as far as I can test.

Further more I tried to install [pijulius/picom](https://github.com/pijulius/picom) manually from source, and in bspwm the problem seems actually same to this repo. So it's just not very suitable for bspwm I guess, or it conflicts with some of my configurations in bspwm, or something else.

Just wondering why I can't find anyone discussed these issues, since I've seen several bspwm rice choose pijulius' picom.

**Comment by allusive-dev** [comment]
Thanks for trying that out.
You shouldn't experience any differences between this repo and pijulius (the difference between this repo and pijulius can be found [here](https://github.com/allusive-dev/picom-allusive#what-is-this))

I don't think the issue is directly with bspwn or qtile but possibly with your configurations of these WM's, considering that it doesn't seem to be a graphics issue.

My only suggestion now is that you both debug your WM configs by either testing if the issue is present on a fresh config or analyzing your code line by line.

**Comment by Poyso** [comment]
I have the same issue, and i run EndavourOS with qtile and running the "original" picom-pijulius

**Comment by allusive-dev** [comment]
> I have the same issue, and i run EndavourOS with qtile and running the "original" picom-pijulius

Thanks for hopping in. At the moment we think the issue may be to do with your user configuration of qtile/bspwm since others have used these WM without issues with picom/pijulius.

You do not need to mention that you are using the original pijulius package because these are the same package like i mentioned here.

> You shouldn't experience any differences between this repo and pijulius (the difference between this repo and pijulius can be found [here](https://github.com/allusive-dev/picom-allusive#what-is-this))

**Comment by Poyso** [comment]
Now i have tried qtile with the default config + the autostart.sh and still returns this problem.

The only thing i noticed,it never happened until i switched between work space (so for me moving the windows to monitor 1 to monitor 2), but might be this behaviour happened even before not sure.

**Comment by allusive-dev** [comment]
At the moment my guess is that this may be an issue with `qtile` and or `bspwm`. I am thinking this because of an issue i recently reviewed which has revealed that workspace switching animation do not play specifically on `dwm`

Unfortunately there is not a lot I can do about this if it is a issue with the WM itself.
But if any other information comes up suggesting otherwise then I hope to be able to try to fix it.

**Comment by AsharAlvany** [comment]
I wanted to comment since I was experiencing the ghost artifacts, as well. I'm also on qtile. From my testing, I was able to get rid of the artifact by turning off shadows. I hope that helps some!

**Comment by allusive-dev** [comment]
Thank you for that. I would close it at that but some have tested with the default config which would be without shadows.

Anyone is welcome to try turning off shadows and let us know how it goes but I'm not sure if that is the issue.

**Comment by vinivosh** [comment]
Can confirm, I am getting these artifacts too with BSPWM and shadows completely turned off.

I also get some weird animations freaking out when I change workspaces. Seems to happen more frequently when I change workspaces after a period of time of not changing them (like > 5 minutes). Instead of animations playing properly once they play out like five times really quickly and sometimes not even using the animation I had configured in `picom.conf` but random ones instead. Pure chaos.

I'm lauching with `picom -b --experimental-backends --backend glx` by the way, if that's relevant.

**Comment by allusive-dev** [comment]
> I'm lauching with `picom -b --experimental-backends --backend glx` by the way, if that's relevant.

From what i can see looking at the source code, experimental backends should only be used if you know what your doing. I would list what experimental-backend does but basically.

If you are not using `dual-kawase` blur or are debugging then do not use it.

 > An idea: can anyone using BSPWM and not getting these issues share their picom.conf and bspwmrc files? So we can test them out.

I agree with this.

**Comment by vinivosh** [comment]
> If you are not using `dual-kawase` blur or are debugging then do not use it.

I am indeed using the `dual-kawase` blur, that's why I'm "stuck" with `--experimental-backends`. It's just so beautiful, haha 😅. A shame that I can't use rounded corners with it…

By the way, I seem to have been able to fix the bug of animations freaking out by setting `animation-clamping` to `true` in the configs. Something about the windows getting bigger than their final size during the animation (and sometimes even parts of them appearing for a brief moment in my other monitor) seems to be what was causing the issue. Much more usable now!

The artifacts for unmapped windows still happens, unfortunately. I thought they had been fixed, because I didn't see any for some 20 minutes of use. But after rebooting the next day they all came back. I did change kernels (`linux-zen` to `linux`) between these reboots, though. I will try again with `linux-zen` later today to see if anything changes.

**Comment by allusive-dev** [comment]
> I am indeed using the `dual-kawase` blur, that's why I'm "stuck" with `--experimental-backends`. It's just so beautiful, haha 😅. A shame that I can't use rounded corners with it…

Wait you can't use rounded corners with that? I am going to test that because that is pretty important.

Yeah i actually found that my animations go insane on XMonad very rarely when i don't use `animation-clamping`.

> I did change kernels (linux-zen to linux) between these reboots, though. I will try again with linux-zen later today to see if anything changes.

I wouldn't know if changing kernels would do anything but let me know if it does.

**Comment by allusive-dev** [comment]
"Rounded corners is only supported on legacy backends"...

Welp. That is something I definitely have to fix.
I'll give it a go today and let you know how i do.

**Comment by allusive-dev** [comment]
Well bad news.
The support for rounded corners is actually not supported in experimental backends and just removing the reset code for the corners does not fix the issue.
I will keep playing with it but I don't know a ton about backends.

**Comment by allusive-dev** [comment]
Hello Everyone,
I wanted to check back in and see if anymore progress has been made and if the issue has been isolated anymore than to a issue with qtile/bspwm?

**Comment by xyzbtw** [comment]
I had the same issue while using openbox, but it got resolved by disabling rounded corners

**Comment by vinivosh** [comment]
I'm going to try this with AwesomeWM tomorrow and see how it goes.

**Comment by SolninjaA** [comment]
I can confirm this also happens on XMonad. It seemed to go away after I open another window, but now that I've switched workspaces and various other things the artifact is always present in the desktop background layer.
![image](https://github.com/allusive-dev/picom-allusive/assets/51935570/c0f09c77-a66d-4272-aa23-b4e62d2eae63)

**Please note, this only happens to me when borders are enabled. Otherwise, this doesn't happen.**

**Comment by allusive-dev** [comment]
Thanks @SolninjaA 
What I can tell is exactly what you said, the issue only seems to happen to people using window borders.

I'm looking at updating picom to the latest upstream version soon, But that may take quite a while so sit tight everyone!

**Comment by SolninjaA** [comment]
Ok, great! It would be great to have picom-allusive on the latest upstream version. I did some testing and you can use "dual-kawase" blur with rounded corners on the upstream version. I'm mentioning that because I noticed that @vinivosh was talking about that.

I haven't tested to see if the artifacts problem is also a problem on the upstream version, but my guess would be no.

**Comment by rudyon** [comment]
on bspw and do not have this issue

**Comment by allusive-dev** [comment]
> on bspw and have this issue aswell

Can I ask whether you are using window borders or not?

**Comment by allusive-dev** [comment]
Hey @vinivosh 
Continuing from what I was saying about hopefully fixing the issue with dual-kawase only being avaliable on experimental-backends which doesn't have rounded-corners.

This issue has been and will be fixed in the update to `yshui/picom-10.2`.
You can check the progress of the update in the [unstable branch](https://github.com/allusive-dev/picom-allusive/tree/unstable).
More information about the status of the update will also be found there.

For everyone. I am hoping that once the update is complete this artifacts issue will be resolved but that is yet to be seen.
Thank you all for your patience!

**Comment by rudyon** [comment]
> > on bspw and have this issue aswell
> 
> Can I ask whether you are using window borders or not?

i was. after disabling window borders it does go away

**Comment by allusive-dev** [comment]
@vinivosh @rudyon @AsharAlvany @Poyso @xyzbtw @draculalaa @Moshe-Shelomov

Everyone, this has been tested by @SolninjaA on `v1.2.1` and they let me know that the issue is now resolved!

If anyone has any more problems to do with this issue please let me know.

**Comment by allusive-dev** [comment]
> That's really great! The issue seems to be indeed resolved 👏
> 
> However, the dual-kawase blur seems to not be working anymore… am I doing something wrong in the config?
> 
> ```
> blur: {
>   # requires: https://github.com/ibhagwan/picom
>   method = "dual_kawase";
>   # method = "kernel";
>   strength = 8;
>   deviation =1.0;
>   # kernel = "11x11gaussian";
>   background = false;
>   background-frame = false;
>   background-fixed = false;
>   kern = "3x3box";
> }
> ```

@vinivosh Please take a look at the [this page](https://github.com/allusive-dev/picom-allusive/wiki/6.-Blur#blur-rule) of the wiki which explains the changes made to blur as of 1.2

**Comment by vinivosh** [comment]
@allusive-dev Ooh, thanks, didn't see the changes in the wiki, my bad.

I like the idea of using a whitelist for blurred windows, much better on performance.

Which makes me think: could we have a whitelist for transparent windows too? I notice some context menus are quite transparent in my setup, even though I never configured anything that resambles that behaviour. (Maybe having both options, blacklisting and whitelisting would be preferable in this case)

**Comment by allusive-dev** [comment]
> Which makes me think: could we have a whitelist for transparent windows too? I notice some context menus are quite transparent in my setup, even though I never configured anything that resambles that behaviour. (Maybe having both options, blacklisting and whitelisting would be preferable in this case)

I'm a bit confused. Doesn't `opacity-rule` already fill that requirement?

**Comment by vinivosh** [comment]
Sounds like it should, but still I see a lot of context menus being made transparent without any explicit rule.

For example:

The discord context menu (I couldn't make flameshot take a screenshot while the right-click menu was open, but I could with this "hover tip", which has the same behaviour)
![image](https://github.com/allusive-dev/picom-allusive/assets/36175157/616082c5-99fe-4912-a547-da5fb58766ea)

The firefox right-click context menu
![image](https://github.com/allusive-dev/picom-allusive/assets/36175157/a255aeef-f660-48bf-b29f-18a3600aeb81)

My opacity rules config:
```bash
# Specify a list of opacity rules, in the format `PERCENT:PATTERN`,
opacity-rule = [
  "70:class_g = 'Xfce4-terminal'",
  "70:class_g = 'kitty'",
  "70:class_g = 'XTerm'",
  "70:class_g = 'xterm'",
  "70:class_g = 'UXTerm'",
  "80:class_g = 'SpeedCrunch'",
  "80:class_g = 'Rofi'",
  # "100:class_g = 'LibreWolf'",
  # "100:class_g = 'Firefox'",
  # "100:class_g = 'Chrome'"
];
```

**Comment by allusive-dev** [comment]
I have a suspicion. @vinivosh 
This sounds like a win-types problem possibly.

If you have a win-types config setup please attach it.

If not try this at the bottom of your config. I'm not sure which one equals what window so I just set them all lol.

```
wintypes:
{
  normal = { opacity = 1; };
  menu = { opacity = 1; };
  tooltip = { opacity = 1; };
  dock = { opacity = 1; };
  dnd = { opacity = 1; };
  popup_menu = { opacity = 1; };
  utility = { opacity = 1; };
  toolbar = { opacity = 1; };
  notification = { opacity = 1; }
};
```

**Comment by allusive-dev** [comment]
Let's move to a new issue. #15

**Comment by allusive-dev** [comment]
For Anyone coming here in the future. This issue was resolved as of update `1.2`
Using any version that matches `1.2` or newer than that will resolve the issue.
